In [1]:
import pandas as pd

In [2]:
purchases = pd.read_csv("../Resources/purchase_data.csv")
purchases.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

First off, we're going to figure out how many unique players have visited this online shop

In [3]:
#Used to count the amount of players
player_count = 0
#Used to store the names of unique players
unique_count = []
#and their gender, for later use
gender = []
#and their age, also for later use
age = []
#everything you see here is from an edit in the future
price = []
#loop through SN column for unique user ID's
for i in range(0, purchases["SN"].count()):
    if purchases["SN"][i] not in unique_count:
        #add user ID's, the gender, and age to their respective lists
        unique_count.append(purchases["SN"][i])
        gender.append(purchases["Gender"][i])
        age.append(purchases["Age"][i])
        price.append(purchases["Price"][i])
        #increment the player count
        player_count +=1
print(player_count)

576


In [4]:
#A dataframe containing the unique players
new = {"Names": unique_count, "Gender": gender, "Age": age, "Price": price}
newdf = pd.DataFrame(new)
newdf

Names  Gender  Age  Price
0          Lisim78    Male   20   3.53
1      Lisovynya38    Male   40   1.56
2       Ithergue48    Male   24   4.88
3    Chamassasya86    Male   24   3.27
4        Iskosia90    Male   23   1.44
..             ...     ...  ...    ...
571         Hala31    Male   21   1.02
572     Jiskjask80    Male   11   4.19
573     Aethedru70  Female   21   3.54
574     Yathecal72    Male   20   3.46
575        Sisur91    Male    7   4.19

[576 rows x 4 columns]

There seem to be a total of 576 unique customer who use the shop.
## Purchasing Analysis (Total)
Now let's clean up the data and figure out:
* The number of unique items in the shop
* The average price of items
* The total number of purchases
* Total revenue

In [5]:
#initiate variables for counting unique items
items = 0
#Because i'm uncreative with names
item_list = []
#loop through the Item Name column for unique item names
for i in range(0, purchases["Item Name"].count()):
    if purchases["Item Name"][i] not in item_list:
        #add item names to the  list, which also increments the items count
        item_list.append(purchases["Item Name"][i])
        items += 1
print(f"We sell {items} items in our shop!")

We sell 179 items in our shop!


In [6]:
#Total Revenue
price_total = purchases["Price"].sum()
print(f"We made ${price_total} today!")

We made $2379.77 today!


In [7]:
#Number of purchases
purch = purchases["Purchase ID"].count()
print(f"We had {purch} visits today!")

We had 780 visits today!


In [8]:
#Average Price
avg_price = round((price_total / purch), 2)
print(f"The average price is ${avg_price}.")

The average price is $3.05.


In [9]:
#Put it all in a neat DataFrame
purchase_analysis = {"Number of Unique Items": [items],
                     "Purchases": [purch],
                     "Total Revenue": [price_total],
                     "Average Price": [avg_price]}
df = pd.DataFrame(purchase_analysis)
df

Number of Unique Items  Purchases  Total Revenue  Average Price
0                     179        780        2379.77           3.05

## Gender Demographics

Now that we got our purchase analysis, next up is checking out the gender demographic.  

In [10]:
#create a dataframe that takes each unique name as a key and the gender as a value
dicts = {"Name": unique_count, "Gender": gender}
df = pd.DataFrame(dicts)
df

Name  Gender
0          Lisim78    Male
1      Lisovynya38    Male
2       Ithergue48    Male
3    Chamassasya86    Male
4        Iskosia90    Male
..             ...     ...
571         Hala31    Male
572     Jiskjask80    Male
573     Aethedru70  Female
574     Yathecal72    Male
575        Sisur91    Male

[576 rows x 2 columns]

In [11]:
#initiate variables
male = 0
female = 0
other = 0
#initiate unimaginative lists
male_list = []
female_list = []
other_list = []
#loop through the Gender column
for i in range(0, df["Gender"].count()):
    if df["Gender"][i] == "Male":#increment the male count
        male_list.append(purchases["Gender"][i])
        male += 1
    elif df["Gender"][i] == "Female":#increment the female count
        female_list.append(purchases["Gender"][i])
        female += 1
    else:#increment the other count
        other_list.append(purchases["Gender"][i])
        other += 1
print(f"There are {male} males, {female} females, and {other} of non-disclosed gender.")


There are 484 males, 81 females, and 11 of non-disclosed gender.


In [12]:
#determine the ratios of each gender 
male_ratio = round((male / player_count * 100), 2)
female_ratio = round((female / player_count * 100), 2)
other_ratio = round((other / player_count * 100), 2)
print(f"The ratios are as follows: {male_ratio}% males, {female_ratio}% females, and {other_ratio}% of undisclosed gender.")

The ratios are as follows: 84.03% males, 14.06% females, and 1.91% of undisclosed gender.


In [13]:
#clean it up as a DataFrame
gender_data = {"Total Count": [male, female, other], 
               "Percentage of Players": [male_ratio, female_ratio, other_ratio]}
gender_df = pd.DataFrame(gender_data, index = ["Male", "Female", "Other / Undisclosed"])
gender_df

Total Count  Percentage of Players
Male                         484                  84.03
Female                        81                  14.06
Other / Undisclosed           11                   1.91

## Purchasing Analysis (Gender)
We've broken down the amount of unique players by gender as well as the ratio of each gender.
Now we're going to analyze the average purchases made by each gender

In [14]:
#reinitiate variables
remale = 0
refemale = 0
reother = 0
#reinitiate unimaginative lists
remale_list = []
refemale_list = []
reother_list = []
#a surprise tool to help us later
male_price = []
female_price = []
other_price = []
#reloop through the Gender column
for i in range(0, purchases["Gender"].count()):
    if purchases["Gender"][i] == "Male":#increment the total male count
        remale_list.append(purchases["Gender"][i])
        male_price.append(purchases["Price"][i])
        remale += 1
    elif purchases["Gender"][i] == "Female":#increment the total female count
        refemale_list.append(purchases["Gender"][i])
        female_price.append(purchases["Price"][i])
        refemale += 1
    else:#increment the total other count
        reother_list.append(purchases["Gender"][i])
        other_price.append(purchases["Price"][i])
        reother += 1
print(f"There are {remale} males, {refemale} females, and {reother} of non-disclosed gender.")


There are 652 males, 113 females, and 15 of non-disclosed gender.


In [15]:
#total purchasing value per gender
male_total_purchase = round(sum(male_price), 2)
female_total_purchase = round(sum(female_price), 2)
other_total_purchase = round(sum(other_price), 2)
print(f"Totals with males having ${male_total_purchase}, females having ${female_total_purchase}, and undisclosed genders with ${other_total_purchase}.")

Totals with males having $1967.64, females having $361.94, and undisclosed genders with $50.19.


In [16]:
#figuring out the average purchasing price per gender
avg_male_purch = round((male_total_purchase / remale), 2)
avg_female_purch = round((female_total_purchase / refemale), 2)
avg_other_purch = round((other_total_purchase / reother), 2)
print(f"An average purchase price of ${avg_male_purch} for males, ${avg_female_purch} for females, and ${avg_other_purch} for undisclosed gender.")

An average purchase price of $3.02 for males, $3.2 for females, and $3.35 for undisclosed gender.


In [17]:
#figuring out the average total purchase per person
avg_total_per_male = round((male_total_purchase / male), 2)
avg_total_per_female = round((female_total_purchase / female), 2)
avg_total_per_other = round((other_total_purchase / other), 2)
print(f"An average total purchase per person of ${avg_total_per_male} for males, ${avg_total_per_female} for females, and ${avg_total_per_other} for undisclosed genders.")

An average total purchase per person of $4.07 for males, $4.47 for females, and $4.56 for undisclosed genders.


In [18]:
#Let's clean it up with a DataFrame
gender_purchase_analysis = {#"Gender": ["Male", "Female", "Other / Undisclosed"],
                            "Purchase Count": [remale, refemale, reother],
                            "Total Purchase Value in Dollars": [male_total_purchase, female_total_purchase, other_total_purchase],
                            "Average Purchase Price in Dollars": [avg_male_purch, avg_female_purch, avg_other_purch],
                            "Average Total Purchase per Person in Dollars": [avg_total_per_male, avg_total_per_female, avg_total_per_other]
                            }
gender_purch_df = pd.DataFrame(gender_purchase_analysis, index = ["Male", "Female", "Other / Undisclosed"])
index = gender_purch_df.index
index.name = "Gender"
gender_purch_df

Purchase Count  Total Purchase Value in Dollars  \
Gender                                                                 
Male                            652                          1967.64   
Female                          113                           361.94   
Other / Undisclosed              15                            50.19   

                     Average Purchase Price in Dollars  \
Gender                                                   
Male                                              3.02   
Female                                            3.20   
Other / Undisclosed                               3.35   

                     Average Total Purchase per Person in Dollars  
Gender                                                             
Male                                                         4.07  
Female                                                       4.47  
Other / Undisclosed                                          4.56

## Age Demographics
Of course, players of many ages are playing Heroes of Pymoli.  So we want to determine the ages of our players, and how many of them take up the playerspace. 

In [19]:
#Separate the Age column and bin certain ages 
binned = [0, 9, 14, 19, 24, 29, 34, 39, 100]
#labels for the bins
lab = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
#cut up the column and create a new "binned" column for the ages
bigbin = pd.cut(newdf["Age"], bins=binned, labels=lab)
bigbin

0      20-24
1        40+
2      20-24
3      20-24
4      20-24
       ...  
571    20-24
572    10-14
573    20-24
574    20-24
575      <10
Name: Age, Length: 576, dtype: category
Categories (8, object): [<10 < 10-14 < 15-19 < 20-24 < 25-29 < 30-34 < 35-39 < 40+]

In [20]:
#simplify the binned column
ages = bigbin
#it is variable time
kid = 0
preteen = 0
teen = 0
young_adult = 0
adult_iguess = 0
adult_2 = 0
adult_3 = 0
old = 0
#loop through the binned column
for i in range(0, ages.count()):
    #for every bin discovered, increment its respective count
    if ages[i] == "<10":
        kid += 1
    elif ages[i] == "10-14":
        preteen += 1
    elif ages[i] == "15-19":
        teen += 1
    elif ages[i] == "20-24":
        young_adult += 1
    elif ages[i] == "25-29":
        adult_iguess += 1
    elif ages[i] == "30-34":
        adult_2 += 1
    elif ages[i] == "35-39":
        adult_3 += 1
    else:
        old += 1
print(f"We've got {kid} under 10, {preteen} people 10-14, {teen} people 15-19, {young_adult} people 20-24, {adult_iguess} people 25-29, {adult_2} people 30-34, {adult_3} people 35-39, and {old} people 40+.")

We've got 17 under 10, 22 people 10-14, 107 people 15-19, 258 people 20-24, 77 people 25-29, 52 people 30-34, 31 people 35-39, and 12 people 40+.


In [21]:
#time for percentages; calculate the percentage of each age category
kid_perc = round((kid / player_count * 100), 2)
preteen_perc = round((preteen / player_count * 100), 2)
teen_perc = round((teen / player_count * 100), 2)
young_adult_perc = round((young_adult / player_count * 100), 2)
adult_iguess_perc = round((adult_iguess / player_count * 100), 2)
adult_2_perc = round((adult_2 / player_count * 100), 2)
adult_3_perc = round((adult_3 / player_count * 100), 2)
old_perc = round((old / player_count * 100), 2)

In [22]:
#And a DataFrame to clean it up
age_demo = {"Total Count": [kid, preteen, teen, young_adult, adult_iguess, adult_2, adult_3, old],
            "Percentage of Players": [kid_perc, preteen_perc, teen_perc, young_adult_perc, adult_iguess_perc, adult_2_perc, adult_3_perc, old_perc]
           }
age_demo_df = pd.DataFrame(age_demo, index = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"])
age_demo_df

Total Count  Percentage of Players
<10             17                   2.95
10-14           22                   3.82
15-19          107                  18.58
20-24          258                  44.79
25-29           77                  13.37
30-34           52                   9.03
35-39           31                   5.38
40+             12                   2.08

## Purchasing Analysis (Age)
With the ages of our players sorted out, let's figure out how much different age groups spend. 

In [31]:
#a dataframe that shows the customers age and how much they spent
ager = purchases["Age"]
purchased = purchases["Price"]
age_purch = pd.DataFrame({"Age": ager, "Purchase": purchased})
age_purch

Age  Purchase
0     20      3.53
1     40      1.56
2     24      4.88
3     24      3.27
4     23      1.44
..   ...       ...
775   21      3.54
776   21      1.63
777   20      3.46
778    7      4.19
779   24      4.60

[780 rows x 2 columns]

In [32]:
#cut up the column and create a new "binned" column for the ages
bigbin2 = pd.cut(ager, bins=binned, labels=lab)

In [34]:
#it is variable time again
akid = 0
apreteen = 0
ateen = 0
ayoung_adult = 0
aadult_iguess = 0
aadult_2 = 0
aadult_3 = 0
aold = 0
#some lists
kid_price = []
preteen_price = []
teen_price = []
young_adult_price = []
adult_iguess_price = []
adult_2_price = []
adult_3_price = []
old_price = []
#loop through the new binned age list
for i in range(0, bigbin2.count()):
    #for every bin discovered, increment its respective count, also 
    if bigbin2[i] == "<10":
        kid_price.append(age_purch["Purchase"][i])
        akid += 1
    elif bigbin2[i] == "10-14":
        preteen_price.append(age_purch["Purchase"][i])
        apreteen += 1
    elif bigbin2[i] == "15-19":
        teen_price.append(age_purch["Purchase"][i])
        ateen += 1
    elif bigbin2[i] == "20-24":
        young_adult_price.append(age_purch["Purchase"][i])
        ayoung_adult += 1
    elif bigbin2[i] == "25-29":
        adult_iguess_price.append(age_purch["Purchase"][i])
        aadult_iguess += 1
    elif bigbin2[i] == "30-34":
        adult_2_price.append(age_purch["Purchase"][i])
        aadult_2 += 1
    elif bigbin2[i] == "35-39":
        adult_3_price.append(age_purch["Purchase"][i])
        aadult_3 += 1
    else:
        old_price.append(age_purch["Purchase"][i])
        aold += 1
print(f"We've got {akid} under 10, {apreteen} people 10-14, {ateen} people 15-19, {ayoung_adult} people 20-24, {aadult_iguess} people 25-29, {aadult_2} people 30-34, {aadult_3} people 35-39, and {aold} people 40+.")

We've got 23 under 10, 28 people 10-14, 136 people 15-19, 365 people 20-24, 101 people 25-29, 73 people 30-34, 41 people 35-39, and 13 people 40+.


In [35]:
#finding the total purchases of each age group
total_kid_purch = round(sum(kid_price), 2)
total_preteen_purch = round(sum(preteen_price), 2)
total_teen_purch = round(sum(teen_price), 2)
total_young_adult_purch = round(sum(young_adult_price), 2)
total_adult_iguess_purch = round(sum(adult_iguess_price), 2)
total_adult_2_purch = round(sum(adult_2_price), 2)
total_adult_3_purch = round(sum(adult_3_price), 2)
total_old_purch = round(sum(old_price), 2)
total_kid_purch

77.13

In [36]:
#figuring out the average purchasing price per person
avg_kid_price = round(total_kid_purch / akid, 2)
avg_preteen_price = round(total_preteen_purch / apreteen, 2)
avg_teen_price = round(total_teen_purch / ateen, 2)
avg_young_adult_price = round(total_young_adult_purch / ayoung_adult, 2)
avg_adult_iguess_price = round(total_adult_iguess_purch / aadult_iguess, 2)
avg_adult_2_price = round(total_adult_2_purch / aadult_2, 2)
avg_adult_3_price = round(total_adult_3_purch / aadult_3, 2)
avg_old_price = round(total_old_purch / aold, 2)
avg_kid_price

3.35

In [38]:
#and now the average total purchase per person
avg_kid_total = round(total_kid_purch / kid, 2)
avg_preteen_total = round(total_preteen_purch / preteen, 2)
avg_teen_total = round(total_teen_purch / teen, 2)
avg_young_adult_total = round(total_young_adult_purch / young_adult, 2)
avg_adult_iguess_total = round(total_adult_iguess_purch / adult_iguess, 2)
avg_adult_2_total = round(total_adult_2_purch / adult_2, 2)
avg_adult_3_total = round(total_adult_3_purch / adult_3, 2)
avg_old_total  = round(total_old_purch / old, 2)

In [58]:
#Time to make a Dataframe with all this in it
ranges = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]
purch_count = [akid, apreteen, ateen, ayoung_adult, aadult_iguess, aadult_2, aadult_3 ,aold]
total_purch = [total_kid_purch,total_preteen_purch, total_teen_purch, total_young_adult_purch, total_adult_iguess_purch, total_adult_2_purch, total_adult_3_purch, total_old_purch] 
avg_price = [avg_kid_price, avg_preteen_price, avg_teen_price, avg_young_adult_price, avg_adult_iguess_price, avg_adult_2_price, avg_adult_3_price,avg_old_price]
avg_total_price = [avg_kid_total, avg_preteen_total, avg_teen_total, avg_young_adult_total, avg_adult_iguess_total, avg_adult_2_total, avg_adult_3_total ,avg_old_total]
pre_df = {"Purchase Count": purch_count, 
          "Total Purchasein Dollars": total_purch, 
          "Average Purchase Price in Dollars": avg_price,
          "Average Total Purchase per Person in Dollars": avg_total_price}
purch_ana = pd.DataFrame(pre_df, index = ranges)
index = purch_ana.index
index.name = "Age Range"
purch_ana


Purchase Count  Total Purchasein Dollars  \
Age Range                                             
<10                    23                     77.13   
10-14                  28                     82.78   
15-19                 136                    412.89   
20-24                 365                   1114.06   
25-29                 101                    293.00   
30-34                  73                    214.00   
35-39                  41                    147.67   
40+                    13                     38.24   

           Average Purchase Price in Dollars  \
Age Range                                      
<10                                     3.35   
10-14                                   2.96   
15-19                                   3.04   
20-24                                   3.05   
25-29                                   2.90   
30-34                                   2.93   
35-39                                   3.60   
40+                                     2.94   

           Average Total Purchase per Person in Dollars  
Age Range                                                
<10                                                4.54  
10-14                                              3.76  
15-19                                              3.86  
20-24                                              4.32  
25-29                                              3.81  
30-34                                              4.12  
35-39                                              4.76  
40+                                                3.19

## Top Spenders
With all these players, wonder whose spending the most?

In [71]:
#me realizing that there is actually a simpler way to do all this
big_spend = purchases.groupby("SN")
#counting the number of purchases done by each customer
how_big_spend = big_spend["Purchase ID"].count()
#calculating the average amount spent per customer
avg_big_spend = round(big_spend["Price"].mean(), 2)
#finding the total purchases made by each customer
total_big_spend = big_spend["Price"].sum()


In [72]:
#Dang, it's already dataframe time
pre_top = {"Names": big_spend,
           "Total Purchases": how_big_spend,
           "Total Spent in Dollars": total_big_spend,
           "Average Spent in Dollars": avg_big_spend}
top_df = pd.DataFrame(pre_top)
#sort it from top five
top_df.sort_values(["Total Spent in Dollars"], ascending=False).head()

Names  \
SN                                                               
Lisosia93    (Lisosia93, [Purchase ID, SN, Age, Gender, Ite...   
Idastidru52      (Idastidru52, [Purchase ID, SN, Age, Gender])   
Chamjask73                (Chamjask73, [Purchase ID, SN, Age])   
Iral74                (Iral74, [Purchase ID, SN, Age, Gender])   
Iskadarya95              (Iskadarya95, [Purchase ID, SN, Age])   

             Total Purchases  Total Spent in Dollars  Average Spent in Dollars  
SN                                                                              
Lisosia93                  5                   18.96                      3.79  
Idastidru52                4                   15.45                      3.86  
Chamjask73                 3                   13.83                      4.61  
Iral74                     4                   13.62                      3.40  
Iskadarya95                3                   13.10                      4.37

## Most Popular Items
What type of items do players love to buy the most?


In [94]:
#grab only the item related columns for this section
items = purchases[["Item ID", "Item Name", "Price"]]
#group them by ID and name
item_grouped = items.groupby(["Item ID", "Item Name"])
#count the amount of times each item was bought
item_purch = item_grouped["Price"].count()
#total up the amount made per item
purch_total = item_grouped["Price"].sum()
#and now individual item costs
each_item = round((purch_total / item_purch), 2)


In [96]:
#dataframe time already?
most_popular_items = pd.DataFrame({"Purchase Count": item_purch, 
                                   "Item Price in Dollars": each_item,
                                   "Total Purchase Value in Dollars": purch_total})
#sort it as top five
most_popular_items.sort_values(["Purchase Count"], ascending=False).head()


Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                      Item Price in Dollars  \
Item ID Item Name                                                             
92      Final Critic                                                   4.61   
178     Oathbreaker, Last Hope of the Breaking Storm                   4.23   
145     Fiery Glass Crusader                                           4.58   
132     Persuasion                                                     3.22   
108     Extraction, Quickblade Of Trembling Hands                      3.53   

                                                      Total Purchase Value in Dollars  
Item ID Item Name                                                                      
92      Final Critic                                                            59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                            50.76  
145     Fiery Glass Crusader                                                    41.22  
132     Persuasion                                                              28.99  
108     Extraction, Quickblade Of Trembling Hands                               31.77

## Most Profitable Items
Well we got all these items in the shop, but what item's making the most money for us?


In [98]:
popular_formatted = most_popular_items.sort_values(["Total Purchase Value in Dollars"], ascending=False).head()
popular_formatted

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                      Item Price in Dollars  \
Item ID Item Name                                                             
92      Final Critic                                                   4.61   
178     Oathbreaker, Last Hope of the Breaking Storm                   4.23   
82      Nirvana                                                        4.90   
145     Fiery Glass Crusader                                           4.58   
103     Singed Scalpel                                                 4.35   

                                                      Total Purchase Value in Dollars  
Item ID Item Name                                                                      
92      Final Critic                                                            59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                            50.76  
82      Nirvana                                                                 44.10  
145     Fiery Glass Crusader                                                    41.22  
103     Singed Scalpel                                                          34.80